keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1215 @@
+   if needed
+
+   Example: I would like to explicitly delegate the restriction of traffic destined
+   for cluster monitoring pods to the developer, allowing them to setup network policy
+   to deny or allow the traffic from/to their application.
+
+There are several unique properties that we need to add in order accomplish the
+user stories above.
+1. Deny rules and, therefore, hierarchical enforcement of policy
+2. Semantics for a cluster-scoped policy object that may include
+   namespaces/workloads that have not been created yet.
+3. Interoperability with existing Kubernetes Network Policy API
+
+### Non-Goals
+
+Our mission is to solve the most common use cases that cluster admins have.
+That is, we don't want to solve for every possible policy permutation a user
+can think of. Instead, we want to design an API that addresses 90-95% use cases
+while keeping the mental model easy to understand and use.
+The focus of this KEP is on cluster scoped controls for east-west traffic within
+a cluster, meaning that an AdminNetworkPolicyPeer is _always_ defined as a set of 
+in cluster objects. Cluster scoped controls for north-south traffic may be addressed via
+future versions of the api resources introduced in this or other future KEPs.
+For the time being, the AdminNetworkPolicy resource introduced by this KEP will
+never affect north-south traffic, and thus also don't override or bypass NetworkPolicies
+with ipBlock rules that select external traffic.
+
+## Proposal
+
+In order to achieve the three primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following resource
+under `policy.networking.k8s.io` API group:
+- AdminNetworkPolicy
+
+### AdminNetworkPolicy resource
+
+An AdminNetworkPolicy (ANP) resource will help the administrators:
+1. Set strict security rules for the cluster, i.e. a developer CANNOT override
+these rules by creating NetworkPolicies that applies to the same workloads as the
+AdminNetworkPolicy does.
+2. Set baseline security rules that describes default connectivity for cluster
+workloads, which CAN be overridden by developer NetworkPolicies if needed.
+
+#### Actions
+
+Unlike the NetworkPolicy resource in which each rule represents an allowed
+traffic, AdminNetworkPolicy will enable administrators to set `Pass`,
+`Deny` or `Allow` as the action of each rule. AdminNetworkPolicy rules should
+be read as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the AdminNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+- Pass: Traffic that matches a `Pass` rule will skip all further rules from all 
+  positive priority ANPs and instead be enforced by the K8s NetworkPolicies. 
+  If there is no K8s NetworkPolicy rule match, and no ANP priority "0" rule 
+  match, traffic will be allowed.
+- Deny: Traffic that matches a `Deny` rule will be dropped. 
+- Allow: Traffic that matches an `Allow` rule will be allowed.
```

Comment:
What about calling this something like "ForceAllow" or similar? Something to call out the fact that it *overrides* any NetworkPolicy?

Reply:
I don't know if "Force" adds enough value in the context?

Reply:
In my mental model, I don't see it conceptually as overriding a network policy - network policy can only allow things anyway, so how do you override an allow with another allow? Rather, it's just an allow rule that gets evaluated earlier in the pipeline.

Reply:
Interesting. My mental model was that ANP is always executed as a layer before NP.

This is because the presence of a distinct "Pass" action implies that NP is always after ANP. I think having it both ways is, therefore, confusing. What happens if we formally define a full ordering (e.g. magic rule 500), and then there is a "Pass" action in a rule that comes after the NP priority? That's counter-intuitive, since this rule will always reject. Would it be possible to catch this case via OpenAPI validation? I don't think so.

In [another comment](https://github.com/kubernetes/enhancements/pull/2522/files/9d7516da561bc2995ed0a589369cec1728a784a7#r795699629) I have a proposal for a fourth action, but it is predicated on ANP always being before NP.
Conversely, if a full ordering is defined, then we can get rid of Pass, right? Pass would essentially be equivalent to `GOTO(500)`.